src/meta-srv/src/region/supervisor.rs

```diff
@@ -0,0 +1,523 @@
+        self.sender.clone()
+    }
+}
+
+impl RegionSupervisor {
+    pub(crate) fn new(
+        options: PhiAccrualFailureDetectorOptions,
+        tick_interval: Duration,
+        selector_context: SelectorContext,
+        selector: SelectorRef,
+        region_migration_manager: RegionMigrationManagerRef,
+        kv_backend: KvBackendRef,
+        peer_lookup: PeerLookupServiceRef,
+    ) -> Self {
+        let (tx, rx) = tokio::sync::mpsc::channel(1024);
+        Self {
+            failure_detector: RegionFailureDetector::new(options),
+            tick_interval,
+            receiver: rx,
+            sender: tx,
+            selector_context,
+            selector,
+            region_migration_manager,
+            kv_backend,
+            peer_lookup,
+        }
+    }
+
+    /// Returns the [`HeartbeatAcceptor`].
+    pub(crate) fn heartbeat_acceptor(&self) -> HeartbeatAcceptor {
+        HeartbeatAcceptor {
+            sender: self.sender.clone(),
+        }
+    }
+
+    /// Returns the [`RegionSupervisorTicker`].
+    pub(crate) fn ticker(&self) -> RegionSupervisorTickerRef {
+        Arc::new(RegionSupervisorTicker {
+            tick_interval: self.tick_interval,
+            sender: self.sender.clone(),
+            tick_handle: Mutex::new(None),
+        })
+    }
+
+    /// Runs the main loop.
+    pub(crate) async fn run(&mut self) {
+        while let Some(event) = self.receiver.recv().await {
+            match event {
+                Event::Tick => {
+                    let regions = self.detect_region_failure();
+                    self.handle_region_failures(regions).await;
+                }
+                Event::HeartbeatArrived(heartbeat) => self.on_heartbeat_arrived(heartbeat),
+                Event::Clear => self.clear(),
+                #[cfg(test)]
+                Event::Dump(sender) => {
+                    let _ = sender.send(self.failure_detector.dump());
+                }
+            }
+        }
+        info!("RegionSupervisor is stopped!");
+    }
+
+    async fn handle_region_failures(&self, mut regions: Vec<(ClusterId, DatanodeId, RegionId)>) {
+        if regions.is_empty() {
+            return;
+        }
+        match self.is_maintenance_mode().await {
+            Ok(false) => {}
+            Ok(true) => {
+                info!("Maintenance mode is enabled, skip failover");
+                return;
+            }
+            Err(err) => {
+                error!(err; "Failed to check maintenance mode");
+                return;
+            }
+        }
+
+        let migrating_regions = regions
+            .extract_if(|(_, _, region_id)| {
+                self.region_migration_manager.tracker().contains(*region_id)
+            })
+            .collect::<Vec<_>>();
+
+        for (cluster_id, datanode_id, region_id) in migrating_regions {
+            self.failure_detector
+                .remove(&(cluster_id, datanode_id, region_id));
+        }
+
+        warn!(
```

Comment:
Maybe also print cluster id so that we can avoid a collect.
```suggestion
        warn!("Detects region failures: {:?}", regions);
```

Reply:
Updated in [e713c21](https://github.com/GreptimeTeam/greptimedb/pull/4172/commits/e713c216c394470df0d8222e749fcd656f745fe6)